---
title: "10.6.21 rmd"
author: "Michelle"
date: "10/6/2021"
output: html_document
---

Fixed ggplot2 so can see x axis better by multiplying BP by 100000... 
```{R, GGPLOT2 CHR1-X median and point plot!!}
library(ggplot2)
DF<- read.csv("~/GitHub/scSeq/GSE158802/DF_10.2.21", row.names=1)
Median <- read.csv("~/GitHub/scSeq/GSE158802/M_10.2.21", row.names=1) 

U=as.factor(unique(DF$Cell))
names(Median)=c(1:37)
Cell_name_DF=data.frame(U,names(Median))
names(Cell_name_DF)=c("original","new")
#altered median names to 1:37 since ggplot code was glitching with all the names for some reason... similar to previous problem with numbers working and X chromosome not working... interesting. 
DF_merged=merge(DF, Cell_name_DF, by.x="Cell", by.y="original")

DF=DF_merged #now the cell names are 1-37! just like in median! 
o=order(as.numeric(DF$new),decreasing=F)
DF=DF[o,]
U=unique(DF$new) ##!!

range(DF$Z.score) #input -4 and 6 as bounds. 

for(i in 1:length(U)){ #had to alter by length to get code to work 
  
  matrix=DF$new==U[i]
  matrix= DF[matrix,]
  matrix[matrix=="23"]="X"
  matrix$CHR <- factor(matrix$CHR, levels=unique(matrix$CHR))


  c=as.numeric(U[i])
  cell_name=Cell_name_DF$original[c]
  title_name=paste("Cell ",U[i],cell_name)
  
  p=ggplot(data = matrix, aes(x = BP*9000000, y = `Z.score`), size=1) +
  geom_point(stat = "identity", size=.1) + labs(x = "Chromosome", y = "Z score")+ylim(-4, 6)
    
  me=Median[,U[i]]
  data_hline=data.frame(CHR=unique(matrix$CHR), hline=me) ###!!!!

      plot_list = list()
  plot_list[[i]] = 
    
   myplot=
      p+ theme_classic()+ theme(axis.text.x = element_blank(),axis.ticks.x = element_blank(), panel.spacing = unit(.5, "lines"), strip.text.x=element_text(size=8)) +ggtitle(title_name)+ 
    facet_grid(~CHR, scales = 'free_x', space = 'free_x', switch = 'x') + 
       
      geom_hline(data=data_hline, aes(yintercept=hline),color="red",size=1) + geom_text(data=data_hline,aes(0,6,label=round(hline,digits=1),fontface="bold.italic"),size=3)
                                                                                         
   #myplot
   
 #print(myplot)}
   
    # Save plots to tiff. Makes a separate file for each plot.
    file_name = paste(title_name, ".tiff", sep="")
    tiff(file_name)
    print(plot_list[[i]])
    dev.off() }

```


ACTION 1: Create centromeres
```{r create centromeres}

# hg19 chromosome sizes
chrom_sizes <- structure(list(V1 = rownames(Median), V2 = c( 249250621L, 
243199373L, 198022430L, 191154276L, 180915260L, 171115067L, 159138663L, 
146364022L, 141213431L, 135534747L, 135006516L, 133851895L, 115169878L, 
107349540L, 102531392L, 90354753L, 81195210L, 78077248L, 59128983L, 
63025520L, 48129895L, 51304566L, 155270560L)), .Names = c("V1", 
"V2"), class = "data.frame", row.names = c(NA, -25L))

# hg19 centromere locations
chromStart = c(121535434L, 92326171L, 90504854L, 49660117L, 46405641L, 58830166L, 58054331L, 43838887L, 47367679L, 39254935L, 51644205L, 34856694L, 16000000L, 16000000L, 17000000L, 35335801L, 22263006L, 15460898L, 24681782L, 26369569L, 11288129L, 13000000L,58632012L) 

chromEnd = c(124535434L, 95326171L, 93504854L, 52660117L, 49405641L, 61830166L, 61054331L, 
46838887L, 50367679L, 42254935L, 54644205L, 37856694L, 19000000L, 19000000L, 20000000L, 38335801L, 25263006L, 18460898L, 27681782L, 29369569L, 14288129L, 16000000L,61632012L)

centromeres <- data.frame(row.names(Median),chromStart,chromEnd)
names(centromeres) = c("chrom", "chromStart", "chromEnd")

```

ACTION 2: Sort and add 'p' or 'q' using if statements on temp1 data... 
```{r use if and then statement to sort through temp1}
temp1<- read.csv("~/GitHub/scSeq/GSE158802/temp1_10.1", row.names=1)

#temp1[temp1$CHR == 23,]$CHR <- "X" #change chr 23 to X... 
#temp1$CHR <- as.factor(temp1$CHR)

#chromosome=unique(temp1$CHR)
chromosome=c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23")
n=nrow(temp1)

gene=data.frame(matrix(ncol=1, nrow=0))
for(i in 1:length(chromosome)){

   matrix=temp1[temp1$CHR == i,] 

   c=(centromeres$chromStart)

      DF_1=data.frame(with(matrix, ifelse(as.numeric(matrix$BP) <= c[i], "p","q")))
            gene=rbind(gene,DF_1)

      }
    
dim(gene)

names(gene)=c("Arm")
new=cbind(gene,temp1)
write.csv(new,file="~/GitHub/scSeq/GSE158802/p_1_temp1_10.6.21")

```

ACTION 3: Identify chromosomes as 1p and 1q, instead of 1. 
```{r}

#run median function now
chr_arm=paste(new$CHR, new$Arm)
new$CHR_ARM=chr_arm
U=unique(new$CHR_ARM)
length(U) #42 but acrocentric (p) chromosomes include 13, 14, 15, 21, 22 
46-5 #41? 

 DF=NULL
group_names
  for(i in 1:length(U)){
            for(j in 1:ncol(L)){

    L=new[new$CHR_ARM== U[i],]
    L=L[,5:41]
    LL=data.frame(median(L[,j]))
    DF=rbind(DF,LL)

  }}

dim(DF) #1554    
1554/42 #42 chr distinctions for 37 cells, perfect!

M=matrix(DF[,1], nrow=42, ncol=37, byrow=TRUE)
M=data.frame(M)
dim(M) #[1] 42 37
names(M)=names(temp1[,4:40])
rownames(M)=U

write.csv(M,file="~/GitHub/scSeq/GSE158802/M_pq10.6.21")


```

ACTION 4: Alter 23 to X 
```{R, time to prep plot convert to gg format}

M_pq10.6 <- read.csv("~/GitHub/scSeq/GSE158802/M_pq10.6.21")
Median=M_pq10.6

p_1_temp1_10.6 <- read.csv("~/GitHub/scSeq/GSE158802/p_1_temp1_10.6.21", row.names=1)
chr_arm=paste(p_1_temp1_10.6$CHR, p_1_temp1_10.6$Arm)

p_1_temp1_10.6=p_1_temp1_10.6[,-3]
p_1_temp1_10.6=p_1_temp1_10.6[,-1]
p_1_temp1_10.6=cbind(chr_arm,p_1_temp1_10.6)


p_1_temp1_10.6[p_1_temp1_10.6$chr_arm == "23 p",]$chr_arm <- "X p" #change chr 23 to X... 
p_1_temp1_10.6[p_1_temp1_10.6$chr_arm == "23 q",]$chr_arm <- "X q" #change chr 23 to X... 

matrix=p_1_temp1_10.6[,4:40]

DF=NULL
for(i in 1:ncol(matrix)){
  M=colnames(matrix)[i]
  temp=cbind(p_1_temp1_10.6[,1:3],matrix[,i],M)
 DF= rbind(DF,temp)

}
colnames(DF)=c("CHR_ARM", "SNP","BP","Z score","Cell")

length(unique(DF$Cell)) #37 perfect! 
length(unique(DF$CHR)) #42 perfect! 

write.csv(DF,file="~/GitHub/scSeq/GSE158802/DF_10.7.21")


```

This time we don't want to plot by bp, but by q and p designation. NOtice this change in the p = ggplot2 code
```{R, copy code 1 but for all p and qs!!}

# remove  21p (one chromosome found apparently)
#there is no 13p or 15p or 22p at all. perfect! 


library(ggplot2)

Median <- read.csv("~/GitHub/scSeq/GSE158802/M_pq10.6.21",row.names=1)
DF= read.csv("~/GitHub/scSeq/GSE158802/DF_10.7.21", row.names=1)

 #remove 21p from all DF datasets :(  SNP is called TPTE
dim(DF) # 213342      5
DF=DF[!(DF$SNP=="TPTE"),]
dim(DF) # 213305      5 
213342-37 #213305 perfect! 

#now remove from median 
dim(Median) #42 37
Median=Median[!(rownames(Median)=="21 p"),]
dim(Median)# 41 37

write.csv(DF,file="~/GitHub/scSeq/GSE158802/DF_10.9.21")
write.csv(Median,file="~/GitHub/scSeq/GSE158802/Median_10.9.21")


U=as.factor(unique(DF$Cell))
names(Median)=c(1:37)
Cell_name_DF=data.frame(U,names(Median))
names(Cell_name_DF)=c("original","new")
#altered median names to 1:37 since ggplot code was glitching with all the names for some reason... similar to previous problem with numbers working and X chromosome not working... interesting. 
DF_merged=merge(DF, Cell_name_DF, by.x="Cell", by.y="original")

DF=DF_merged #now the cell names are 1-37! just like in median! 
o=order(as.numeric(DF$new),decreasing=F)
DF=DF[o,]
U=unique(DF$new) ##!!


for(i in 1:length(U)){ #had to alter by length to get code to work 
  
  matrix=DF$new==U[i]
  matrix= DF[matrix,]
  matrix[matrix=="23"]="X"
  matrix$CHR_ARM <- factor(matrix$CHR_ARM, levels=unique(matrix$CHR_ARM))


  c=as.numeric(U[i])
  cell_name=Cell_name_DF$original[c]
  title_name=paste("Cell ",U[i],cell_name)
  
  p=ggplot(data = matrix, aes(x = CHR_ARM, y = Z.score),ylim=c(-4,6)) +
 
  geom_dotplot(binaxis='y',stackdir='centerwhole',stackratio=.5,dotsize=.1,width=5) 

    
  me=Median[,U[i]]
  data_hline=data.frame(CHR_ARM=unique(matrix$CHR_ARM), hline=me) ###!!!!

      plot_list = list()
  plot_list[[i]] = 
    
    #myplot=
    
    p+ ggtitle(title_name)+ geom_point(data=data_hline,mapping=aes(x=CHR_ARM,y=hline,size=hline,color='red')) + scale_size_continuous(breaks = 1:4)+ theme(axis.title.x = element_text()) + theme(legend.position="none")+theme(axis.title.x = element_text(margin = unit(c(10, 0, 0, 0), "mm")))+ 
 geom_text(data=data_hline,aes(CHR_ARM,-4,label=round(hline,digits=1),fontface="bold.italic"),size=3) 
    
     #print(myplot)}


    # Save plots to tiff. Makes a separate file for each plot.
    file_name = paste(title_name, ".tiff", sep="")
    png(file_name,width=1000, height=600)
    print(plot_list[[i]])
    dev.off() 
}





```

```{R, plot medians of all chromosomes p through q}

DF <- read.csv("~/GitHub/scSeq/GSE158802/DF_10.9.21", row.names=1)
Median <- read.csv("~/GitHub/scSeq/GSE158802/Median_10.9.21", row.names=1)


temp1=DF
CHR_ARM=paste(temp1$CHR, temp1$Arm)
temp1=cbind(CHR_ARM,temp1)
U=unique(temp1$CHR_ARM)

matrix=Median 
names(matrix)=c(1:37)

  for(i in 1:length(U)){
  m=matrix[i,]
  title_name=paste("Chromosome",U[i])

  
  plot_list = list()
  plot_list[[i]] = boxplot(m,ylim=c(-1,1),xlab="cell number", ylab="Z score", main=title_name)
    
    file_name = paste(title_name, ".tiff", sep="")
    tiff(file_name,width=1000,height=600)
    print(plot_list[[i]])
    dev.off()
    }

#for some reason isn't saving... even though same code for ggplot2 works :( 
  

```









```{R, extra code}
library(ggplot2)
Median <- read.csv("~/GitHub/scSeq/GSE158802/M_pq10.6.21",row.names=1)
DF= read.csv("~/GitHub/scSeq/GSE158802/DF_10.7.21")

U=as.factor(unique(DF$Cell))
names(Median)=c(1:37)
Cell_name_DF=data.frame(U,names(Median))
names(Cell_name_DF)=c("original","new")
#altered median names to 1:37 since ggplot code was glitching with all the names for some reason... similar to previous problem with numbers working and X chromosome not working... interesting. 
DF_merged=merge(DF, Cell_name_DF, by.x="Cell", by.y="original")

DF=DF_merged #now the cell names are 1-37! just like in median! 
o=order(as.numeric(DF$new),decreasing=F)
DF=DF[o,]

library(ggplot2)

U=unique(DF$new)

for(i in 1:length(U)){ #had to alter by length to get code to work 
  
  matrix=DF$new==U[i]
  matrix= DF[matrix,]
  c=as.numeric(U[i])
  cell_name=Cell_name_DF$original[c]
  title_name=paste("Cell",i,cell_name)
  
  
p=ggplot(data = matrix, aes(x = CHR_ARM, y = Z.score), size=2) +
geom_point(stat = "identity",size=.1) + labs(x = "Chromosome", y = "Z score")
  

  #geom_dotplot(binaxis='y',stackdir='center',dotsize=.2)
  
  me=Median[,U[i]]
  data_hline=data.frame(CHR_ARM=unique(matrix$CHR_ARM), hline=me) ###!!!!
  plot_list = list()
  plot_list[[i]] = 

 myplot= p+ theme_classic()+ theme(axis.text.x = element_blank(),axis.ticks.x = element_blank(), panel.spacing = unit(.5, "lines"), strip.text.x=element_text(size=10),strip.placement='inside') +ggtitle(title_name)  +    
    
    #facet_grid(~CHR_ARM, scales = 'free_y', space = 'free_x', switch = 'x')
    facet_grid(~CHR_ARM)+ geom_hline(data=data_hline, aes(yintercept=hline),color="red", size=3) + geom_text(data=data_hline,aes(0,3.5,label=round(hline,digits=1), vjust=-1,fontface="bold.italic"))
     
    # Save plots to tiff. Makes a separate file for each plot.
    file_name = paste(title_name, ".tiff", sep="")
    tiff(file_name,res=100 )
    print(plot_list[[i]])
    dev.off() }

```
